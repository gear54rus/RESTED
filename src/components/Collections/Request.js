import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';
import { ListGroup } from 'react-bootstrap';
import flow from 'lodash.flow';

import IconButton from 'components/IconButton';
import * as CollectionActions from 'store/collections/actions';
import * as RequestActions from 'store/request/actions';
import * as ConfigActions from 'store/config/actions';
import { isDefaultCompact } from 'store/options/selectors';
import { getEditingRequest } from 'store/config/selectors';
import selectText from 'utils/selectText';

import { StyledRequest, RequestButtons, MainContentDiv } from './StyledComponents';
import * as Type from './dropTypes';

/**
 * Specifies the drag source contract.
 */
const requestSource = {
  canDrag(props) {
    // Disallow drag if editing request
    return !props.isEditing;
  },

  beginDrag({ request, index, collectionIndex }) {
    return { request, index, collectionIndex };
  },
};

const requestTarget = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dragCollectionIndex = monitor.getItem().collectionIndex;
    const dragUUID = monitor.getItem().request.id;
    const hoverIndex = props.index;
    const hoverCollectionIndex = props.collectionIndex;
    const hoverUUID = props.request.id;

    // Abort further processing if dragged item is hovering over itself
    if (dragUUID === hoverUUID) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().index = hoverIndex;
    monitor.getItem().collectionIndex = hoverCollectionIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderRequest({
      // Source
      collectionIndex: dragCollectionIndex,
      requestIndex: dragIndex,
    }, {
      // Target
      collectionIndex: hoverCollectionIndex,
      requestIndex: hoverIndex,
    });
  },
};

class Request extends React.Component {
  static propTypes = {
    connectDragSource: PropTypes.func.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
    isDragging: PropTypes.bool.isRequired,
    index: PropTypes.number.isRequired,
    collectionIndex: PropTypes.number.isRequired,
    renameRequest: PropTypes.func.isRequired,
    selectRequest: PropTypes.func.isRequired,
    sendRequest: PropTypes.func.isRequired,
    deleteRequest: PropTypes.func.isRequired,
    toggleEditMode: PropTypes.func.isRequired,
    isEditing: PropTypes.bool.isRequired,
    defaultCompact: PropTypes.bool.isRequired,
    request: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string,
      method: PropTypes.string.isRequired,
      url: PropTypes.string.isRequired,
    }).isRequired,
  };

  state = {
    compact: this.props.defaultCompact,
  };

  toggleCompact = () => {
    this.setState({ compact: !this.state.compact });
  }

  toggleEdit = () => {
    this.props.toggleEditMode(this.props.request);
  }

  renameRequest = e => {
    const { collectionIndex, index, renameRequest } = this.props;
    e.preventDefault();

    this.toggleEdit();
    renameRequest(collectionIndex, index, this.nameRef.value);
  }

  renderRequest() {
    const { compact } = this.state;
    const { request, index, collectionIndex, isEditing } = this.props;
    const { method, name, url } = request;

    return (
      <div>
        {!compact && <h4>{method}</h4>}
        {isEditing ? (
          <form onSubmit={this.renameRequest}>
            <label
              className="sr-only"
              htmlFor={`${collectionIndex}.${index}.RequestName`}
            >
              Request name
            </label>
            <input
              autoFocus
              onFocus={selectText}
              id={`${collectionIndex}.${index}.RequestName`}
              defaultValue={name}
              ref={ref => { this.nameRef = ref; }}
            />
            <IconButton
              tooltip="Save"
              icon="check"
            />
          </form>
        ) : (
          name || url
        )}
      </div>
    );
  }

  render() {
    const { compact } = this.state;
    const {
      connectDragSource,
      connectDropTarget,
      isDragging,
      request,
      collectionIndex,
      selectRequest,
      sendRequest,
      deleteRequest,
      isEditing,
    } = this.props;

    return connectDragSource(connectDropTarget(
      <div>{/* Need a wrapper div for React DnD support */}
        <StyledRequest isDragging={isDragging}>
          <ListGroup componentClass="div">
            <div className="list-group-item">
              <RequestButtons compact={compact}>
                {!compact && (
                  <IconButton
                    tooltip="Toggle edit"
                    icon="pencil-alt"
                    onClick={this.toggleEdit}
                  />
                )}
                {compact && (
                  <IconButton
                    tooltip="Expand"
                    icon="plus"
                    onClick={this.toggleCompact}
                  />
                )}

                {!compact && !isEditing && (
                  <IconButton
                    tooltip="Minimize"
                    icon="minus"
                    onClick={this.toggleCompact}
                  />
                )}
                {!compact && isEditing && (
                  <IconButton
                    tooltip="Delete"
                    icon="trash"
                    onClick={() => deleteRequest(request.id, collectionIndex)}
                  />
                )}
              </RequestButtons>

              {isEditing ? (
                <MainContentDiv compact={compact}>
                  {this.renderRequest()}
                </MainContentDiv>
              ) : (
                <MainContentDiv
                  compact={compact}
                  onClick={() => selectRequest(request)}
                  onDoubleClick={() => sendRequest(request)}
                >
                  {this.renderRequest()}
                </MainContentDiv>
              )}
            </div>
          </ListGroup>
        </StyledRequest>
      </div>,
    ));
  }
}

const mapStateToProps = (state, props) => ({
  isEditing: getEditingRequest(state)
    ? getEditingRequest(state).id === props.request.id
    : false,
  defaultCompact: isDefaultCompact(state),
});

export default flow(
  DropTarget(Type.Request, requestTarget, connector => ({
    connectDropTarget: connector.dropTarget(),
  })),
  DragSource(Type.Request, requestSource, (connector, monitor) => ({
    connectDragSource: connector.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  connect(mapStateToProps, {
    ...CollectionActions,
    ...RequestActions,
    ...ConfigActions,
  }),
)(Request);

