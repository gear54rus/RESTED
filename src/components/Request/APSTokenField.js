import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
// import { fieldPropTypes } from 'redux-form'; for input proptypes
import { Col, FormGroup, FormControl, Checkbox } from 'react-bootstrap';

import Collapsible from 'components/Collapsible';
import { getTokenDescription, getLastTokenChangeTime } from 'store/aps/selectors';

import { SmallHelpBlock } from './StyledComponents';

function APSTokenField({ apsToken, description, lastFetchTime }) {
  return (
    <Collapsible
      title="APS Token"
      id="apsToken"
    >
      <FormGroup>
        <Col xs={10}>
          <FormControl
            type="text"
            placeholder="APS Token in base64 format"
            {...apsToken.value.input}
          />
          {description && (
            <SmallHelpBlock>
              {description} fetched at {`${lastFetchTime.getHours()}:${lastFetchTime.getMinutes()}`}
            </SmallHelpBlock>
          )}
        </Col>

        <Col xs={2}>
          <Checkbox
            {...apsToken.send.input}
            checked={apsToken.send.input.value}
          >
            Send APS Token?
          </Checkbox>
        </Col>
      </FormGroup>
    </Collapsible>
  );
}

APSTokenField.propTypes = {
  apsToken: PropTypes.shape({
    value: PropTypes.shape({}).isRequired, // onDragStart is not passed, causes error
    send: PropTypes.shape({}).isRequired, // onDragStart is not passed, causes error
  }).isRequired,
  description: PropTypes.string,
  lastFetchTime: PropTypes.instanceOf(Date),
};

const mapStateToProps = state => ({
  description: getTokenDescription(state),
  lastFetchTime: getLastTokenChangeTime(state),
});

export default connect(mapStateToProps)(APSTokenField);
