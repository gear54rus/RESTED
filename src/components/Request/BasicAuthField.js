import React, { PropTypes } from 'react';
import { Col, FormGroup, FormControl, Checkbox } from 'react-bootstrap';

import Collapsible from 'components/Collapsible';

export class BasicAuthField extends React.Component {
  static propTypes = {
    basicAuth: PropTypes.shape({
      username: PropTypes.shape({
        input: PropTypes.shape({}).isRequired,
      }).isRequired,
      password: PropTypes.shape({
        input: PropTypes.shape({}).isRequired,
      }).isRequired,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    this.toggleShowPassword = this.toggleShowPassword.bind(this);
  }

  state = {
    showPassword: false,
  };

  toggleShowPassword(e) {
    this.setState({
      showPassword: e.target.checked,
    });
  }

  render() {
    const { basicAuth } = this.props;

    return (
      <Collapsible
        title="Basic auth"
        id="basicAuth"
      >
        <FormGroup controlId="method">
          <Col xs={5}>
            <FormControl
              type="text"
              placeholder="Username"
              {...basicAuth.username.input}
            />
          </Col>
          <Col xs={5}>
            <FormControl
              type={this.state.showPassword ? 'text' : 'password'}
              placeholder="Password"
              {...basicAuth.password.input}
            />
          </Col>
          <Col xs={2}>
            <Checkbox
              checked={this.state.showPassword}
              onChange={this.toggleShowPassword}
            >
              Show password?
            </Checkbox>
          </Col>
        </FormGroup>
      </Collapsible>
    );
  }
}

export default BasicAuthField;

