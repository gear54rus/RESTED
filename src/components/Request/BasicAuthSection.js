import React, { PropTypes } from 'react';
import { Col, Row, FormControl, Checkbox } from 'react-bootstrap';
// import { fieldPropTypes } from 'redux-form'; for inputs

class BasicAuthSection extends React.Component {
  static propTypes = {
    username: PropTypes.shape({}).isRequired, // onDragStart is not passed, causes error
    password: PropTypes.shape({}).isRequired, // onDragStart is not passed, causes error
    small: PropTypes.bool,
  };

  state = {
    showPassword: false,
  };

  toggleShowPassword(e) {
    this.setState({
      showPassword: e.target.checked,
    });
  }

  render() {
    const {
      username,
      password,
      small,
    } = this.props;

    return (
      <Row>
        <Col
          xs={small ? 4 : 5}
        >
          <FormControl
            type="text"
            placeholder="Username"
            {...username.input}
          />
        </Col>

        <Col
          xs={5}
        >
          <FormControl
            type={this.state.showPassword ? 'text' : 'password'}
            placeholder="Password"
            {...password.input}
          />
        </Col>

        <Col
          xs={small ? 3 : 2}
        >
          <Checkbox
            checked={this.state.showPassword}
            onChange={e => this.toggleShowPassword(e)}
          >
            Show password?
          </Checkbox>
        </Col>
      </Row>
    );
  }
}

export default BasicAuthSection;
