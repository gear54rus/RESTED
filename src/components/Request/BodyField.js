import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Field, FieldArray } from 'redux-form';
import { FormGroup, FormControl, Row, Col, ControlLabel } from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsible from 'components/Collapsible';
import IconButton from 'components/IconButton';
import * as RequestActions from 'store/request/actions';
import { getBodyType } from 'store/request/selectors';

import { UnstyledButton, FormDataFields } from './StyledComponents';

function renderField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function renderFormDataFields(props) {
  const { fields, meta } = props;
  return (
    <FormDataFields>
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`requestBodyGroup${key}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={renderField}
              placeholder="Name"
            />
          </Col>
          <Col xs={5}>
            <Field
              name={`${field}.value`}
              component={renderField}
              placeholder="Value"
            />
          </Col>
          <Col xs={2}>
            <IconButton
              id={`removeFormDataButton${key}`}
              tooltip="Remove form field"
              icon="trash"
              onClick={() => fields.remove(key)}
            />
          </Col>
        </FormGroup>
      ))}
      <Row>
        <Col xs={12}>
          <UnstyledButton
            id="addParameter"
            bsStyle="link"
            onClick={() => fields.push({})}
          >
            <Fonticon icon="plus" />
            Add parameter
          </UnstyledButton>
        </Col>
      </Row>
    </FormDataFields>
  );
}

/* eslint-disable react/no-unused-prop-types */
renderFormDataFields.propTypes = {
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

function renderSingleField({ input }) {
  return (
    <Col xs={12}>
      <FormGroup controlId="requestBody">
        <FormControl
          componentClass="textarea"
          rows={10}
          {...input}
        />
      </FormGroup>
    </Col>
  );
}

renderSingleField.propTypes = {
  input: PropTypes.shape({}).isRequired,
};

function renderBodyType({ input, changeBodyType }) {
  return (
    <FormGroup controlId="bodyType">
      <Col componentClass={ControlLabel} xs={2}>
        Type
      </Col>

      <Col xs={8}>
        <FormControl
          componentClass="select"
          placeholder="Body type"
          {...input}
          onChange={newVal => {
            changeBodyType(newVal.target.value);
            input.onChange(newVal);
          }}
        >
          <option value="custom">Custom</option>
          <option value="json">JSON</option>
          <option value="multipart">Multipart form data</option>
          <option value="urlencoded">URLencoded form data</option>
        </FormControl>
      </Col>
    </FormGroup>
  );
}

renderBodyType.propTypes = {
  input: PropTypes.shape({}).isRequired,
  changeBodyType: PropTypes.func.isRequired,
};

function renderDataField(type) {
  switch (type) {
    case 'json':
    case 'multipart':
    case 'urlencoded':
      return (
        <FieldArray
          name="formData"
          component={renderFormDataFields}
        />
      );
    default:
      return (
        <Field
          name="data"
          component={renderSingleField}
        />
      );
  }
}

export function BodyField({ bodyType, changeBodyType }) {
  return (
    <Collapsible title="Request body" id="requestBody">
      <Field
        name="bodyType"
        component={renderBodyType}
        changeBodyType={changeBodyType}
      />
      {renderDataField(bodyType)}
    </Collapsible>
  );
}

BodyField.defaultProps = {
  bodyType: 'json',
};

BodyField.propTypes = {
  bodyType: PropTypes.oneOf(['json', 'multipart', 'urlencoded', 'custom']),
  changeBodyType: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  bodyType: getBodyType(state),
});

export default connect(mapStateToProps, RequestActions)(BodyField);

