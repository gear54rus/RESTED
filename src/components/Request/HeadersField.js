import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import { FormGroup, FormControl, Row, Col } from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsible from 'components/Collapsible';

import HeaderNameAutosuggest from './HeaderNameAutosuggest';
import { UnstyledButton, TrashButton } from './StyledComponents';

function renderValueField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderValueField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function HeadersField({ meta, fields }) {
  return (
    <Collapsible
      title="Headers"
      id="headers"
    >
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`header.${field}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={HeaderNameAutosuggest}
              placeholder="Name"
            />
          </Col>
          <Col xs={5}>
            <Field
              name={`${field}.value`}
              component={renderValueField}
              placeholder="Value"
            />
          </Col>
          <Col xs={2}>
            <TrashButton
              id={`removeHeaderButton${key}`}
              tooltip="Remove header"
              icon="trash"
              onClick={e => {
                e.preventDefault();
                fields.remove(key);
              }}
            />
          </Col>
        </FormGroup>
      ))}

      <Row>
        <Col xs={12}>
          <UnstyledButton
            id="addHeaderButton"
            bsStyle="link"
            onClick={() => fields.push({})}
          >
            <Fonticon icon="plus" />
            Add header
          </UnstyledButton>
        </Col>
      </Row>
    </Collapsible>
  );
}

// ESLint is not smart enough to see that these are used
/* eslint-disable react/no-unused-prop-types */
HeadersField.propTypes = {
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

export default HeadersField;

