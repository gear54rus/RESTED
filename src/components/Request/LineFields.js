import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Col, FormGroup, Button, DropdownButton, FormControl, InputGroup, MenuItem } from 'react-bootstrap';
import { Typeahead } from 'react-bootstrap-typeahead';

import { REQUEST_METHODS } from 'constants/constants';
import { fieldNoDragShape } from 'propTypes/field';
import { isEditMode } from 'store/config/selectors';
import * as actions from 'store/request/actions';

function LineFields({ method, url, placeholderUrl, copyCurl, editMode }) {
  const btnCaption = editMode ? 'Update request' : 'Send request';
  const btnStyle = editMode ? 'success' : 'primary';

  return (
    <FormGroup>
      <Col sm={2}>
        <Typeahead
          options={REQUEST_METHODS}
          filterBy={() => true}
          {...method.input}
          onChange={selected => method.input.onChange(selected[0])}
          onInputChange={value => method.input.onChange(value)}
          selected={[method.input.value]}
        />
      </Col>
      <Col sm={10}>
        <InputGroup>
          <FormControl
            type="text"
            placeholder={placeholderUrl}
            autoFocus
            {...url.input}
          />
          <InputGroup.Button>
            <Button
              bsStyle={btnStyle}
              type="submit"
            >
              {btnCaption}
            </Button>
            <DropdownButton
              bsStyle={btnStyle}
              pullRight
              title=""
              id="request-copy-curl"
            >
              <MenuItem
                eventKey="1"
                onClick={copyCurl}
              >
                Copy cURL
              </MenuItem>
            </DropdownButton>
          </InputGroup.Button>
        </InputGroup>
      </Col>
    </FormGroup>
  );
}

LineFields.propTypes = {
  method: PropTypes.shape(fieldNoDragShape).isRequired,
  url: PropTypes.shape(fieldNoDragShape).isRequired,
  placeholderUrl: PropTypes.string.isRequired,
  copyCurl: PropTypes.func.isRequired,
  editMode: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  placeholderUrl: state.request.placeholderUrl,
  editMode: isEditMode(state),
});

export { LineFields };
export default connect(mapStateToProps, { copyCurl: actions.copyCurl })(LineFields);
