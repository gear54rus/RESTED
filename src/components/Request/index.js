import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { reduxForm, Fields, FieldArray, FormSection, getFormValues } from 'redux-form';
import { Panel, Form } from 'react-bootstrap';
import flow from 'lodash.flow';

import * as requestActions from 'store/request/actions';
import * as collectionsActions from 'store/collections/actions';
import { isEditMode } from 'store/config/selectors';
import { DEFAULT_REQUEST } from 'constants/constants';
import AuthField from 'components/Authentication';

import Titlebar from './Titlebar';
import LineFields from './LineFields';
import HeadersField from './HeadersField';
import BodyField from './BodyField';

export const requestForm = 'request';

function Request(props) {
  const {
    formValues = {},
    handleSubmit,
    editMode,
    sendRequest,
    updateRequest,
  } = props;

  return (
    <Panel>
      <Panel.Heading>
        <Panel.Title componentClass={Titlebar} />
      </Panel.Heading>
      <Panel.Body>
        <Form
          horizontal
          onSubmit={handleSubmit(editMode ? updateRequest : sendRequest)}
        >
          <Fields
            names={['method', 'url']}
            component={LineFields}
          />
          <FieldArray
            name="headers"
            component={HeadersField}
          />
          <FormSection
            name="auth"
            component={AuthField}
          />
          {!['GET', 'HEAD'].includes(formValues.method) && (
            <BodyField />
          )}
        </Form>
      </Panel.Body>
    </Panel>
  );
}

Request.propTypes = {
  formValues: PropTypes.shape({}),
  handleSubmit: PropTypes.func.isRequired,
  editMode: PropTypes.bool.isRequired,
  sendRequest: PropTypes.func.isRequired,
  updateRequest: PropTypes.func.isRequired,
};

const formOptions = {
  form: requestForm,
};

const mapStateToProps = state => ({
  useFormData: state.request.useFormData,
  initialValues: DEFAULT_REQUEST,
  formValues: getFormValues(requestForm)(state),
  editMode: isEditMode(state),
});

export { Request };
export default flow(
  reduxForm(formOptions),
  connect(mapStateToProps, {
    ...requestActions,
    ...collectionsActions,
  }),
)(Request);
