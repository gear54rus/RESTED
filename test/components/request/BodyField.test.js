import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider, connect } from 'react-redux';
import { reduxForm } from 'redux-form';

/* eslint-disable import/no-unresolved */
import { BodyField } from 'components/Request/BodyField';
import makeStore from '../../makeStore';

describe('BodyField', () => {
  const store = makeStore({});

  it('should match the previous snapshot when !useFormData', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="custom"
        changeBodyType={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should match the previous snapshot when bodyType is json', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="json"
        changeBodyType={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should match the previous snapshot when bodyType is urlencoded', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="urlencoded"
        changeBodyType={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should match the previous snapshot when bodyType is multipart', () => {
    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="multipart"
        changeBodyType={() => {}}
      />
    ));

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render the formData fields', () => {
    const initialValues = {
      formData: [
        {
          name: 'foo',
          value: 'something',
        }, {
          name: 'bar',
          value: 'something2',
        },
      ],
    };

    let Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="json"
        changeBodyType={() => {}}
      />
    ));

    Decorated = connect(() => ({
      initialValues,
    }))(Decorated);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    // Filter out the initial checkbox
    const inputs = tree
      .find('input')
      .filterWhere(i => i.prop('type') !== 'checkbox');

    expect(inputs.length).toBe(4);

    let i = 0;
    initialValues.formData.forEach(header => {
      const nameField = inputs.at(i++); // eslint-disable-line no-plusplus
      const valField = inputs.at(i++); // eslint-disable-line no-plusplus

      expect(nameField.prop('placeholder')).toBe('Name');
      expect(valField.prop('placeholder')).toBe('Value');
      expect(nameField.prop('value')).toBe(header.name);
      expect(valField.prop('value')).toBe(header.value);
    });
  });

  it('should render the data textarea', () => {
    const initialValues = {
      data: 'foo crux bluxorb',
    };

    let Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="custom"
        changeBodyType={() => {}}
      />
    ));

    Decorated = connect(() => ({
      initialValues,
    }))(Decorated);

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    const input = tree.find('textarea');

    expect(input.length).toBe(1);
  });

  it('should handle chaning the selected body type', () => {
    const changeBodyType = jest.fn();

    const Decorated = reduxForm({
      form: 'testForm',
    })(() => (
      <BodyField
        bodyType="json"
        changeBodyType={changeBodyType}
      />
    ));

    const tree = mount(
      <Provider store={store}>
        <Decorated />
      </Provider>,
    );

    expect(changeBodyType).not.toHaveBeenCalled();

    const select = tree.find('select').first();
    select.simulate('change', { target: { value: 'multipart' } });

    expect(changeBodyType).toHaveBeenCalledWith('multipart');

    select.simulate('change', { target: { value: 'custom' } });
    expect(changeBodyType).toHaveBeenCalledWith('custom');
  });
});

