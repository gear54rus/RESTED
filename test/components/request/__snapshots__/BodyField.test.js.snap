// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyField should match the previous snapshot when !useFormData 1`] = `
<div>
  <button
    className="diHGTo btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4
      className="dQgEds"
    >
      Request body
      <i
        className="fa fa-angle-right fWoLNy"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <label
          className="col-xs-2 control-label"
          htmlFor="bodyType"
        >
          Type
        </label>
        <div
          className="col-xs-8"
        >
          <select
            className="form-control"
            id="bodyType"
            name="bodyType"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Body type"
            type={undefined}
            value=""
          >
            <option
              value="custom"
            >
              Custom
            </option>
            <option
              value="json"
            >
              JSON
            </option>
            <option
              value="multipart"
            >
              Multipart form data
            </option>
            <option
              value="urlencoded"
            >
              URLencoded form data
            </option>
          </select>
        </div>
      </div>
      <div
        className="col-xs-12"
      >
        <div
          className="form-group"
        >
          <textarea
            className="form-control"
            id="requestBody"
            name="data"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            rows={10}
            type={undefined}
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BodyField should match the previous snapshot when bodyType is json 1`] = `
<div>
  <button
    className="diHGTo btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4
      className="dQgEds"
    >
      Request body
      <i
        className="fa fa-angle-right fWoLNy"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <label
          className="col-xs-2 control-label"
          htmlFor="bodyType"
        >
          Type
        </label>
        <div
          className="col-xs-8"
        >
          <select
            className="form-control"
            id="bodyType"
            name="bodyType"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Body type"
            type={undefined}
            value=""
          >
            <option
              value="custom"
            >
              Custom
            </option>
            <option
              value="json"
            >
              JSON
            </option>
            <option
              value="multipart"
            >
              Multipart form data
            </option>
            <option
              value="urlencoded"
            >
              URLencoded form data
            </option>
          </select>
        </div>
      </div>
      <div
        className="eWjddY"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <button
              className="hvMBRh btn btn-link"
              disabled={false}
              id="addParameter"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-plus"
                role="presentation"
              />
              Add parameter
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BodyField should match the previous snapshot when bodyType is multipart 1`] = `
<div>
  <button
    className="diHGTo btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4
      className="dQgEds"
    >
      Request body
      <i
        className="fa fa-angle-right fWoLNy"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <label
          className="col-xs-2 control-label"
          htmlFor="bodyType"
        >
          Type
        </label>
        <div
          className="col-xs-8"
        >
          <select
            className="form-control"
            id="bodyType"
            name="bodyType"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Body type"
            type={undefined}
            value=""
          >
            <option
              value="custom"
            >
              Custom
            </option>
            <option
              value="json"
            >
              JSON
            </option>
            <option
              value="multipart"
            >
              Multipart form data
            </option>
            <option
              value="urlencoded"
            >
              URLencoded form data
            </option>
          </select>
        </div>
      </div>
      <div
        className="eWjddY"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <button
              className="hvMBRh btn btn-link"
              disabled={false}
              id="addParameter"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-plus"
                role="presentation"
              />
              Add parameter
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BodyField should match the previous snapshot when bodyType is urlencoded 1`] = `
<div>
  <button
    className="diHGTo btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <h4
      className="dQgEds"
    >
      Request body
      <i
        className="fa fa-angle-right fWoLNy"
        role="presentation"
      />
    </h4>
  </button>
  <div
    aria-expanded={null}
    className="collapse row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="form-group"
      >
        <label
          className="col-xs-2 control-label"
          htmlFor="bodyType"
        >
          Type
        </label>
        <div
          className="col-xs-8"
        >
          <select
            className="form-control"
            id="bodyType"
            name="bodyType"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Body type"
            type={undefined}
            value=""
          >
            <option
              value="custom"
            >
              Custom
            </option>
            <option
              value="json"
            >
              JSON
            </option>
            <option
              value="multipart"
            >
              Multipart form data
            </option>
            <option
              value="urlencoded"
            >
              URLencoded form data
            </option>
          </select>
        </div>
      </div>
      <div
        className="eWjddY"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <button
              className="hvMBRh btn btn-link"
              disabled={false}
              id="addParameter"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-plus"
                role="presentation"
              />
              Add parameter
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
